![alt text](https://i.imgur.com/1WaY7aA.png)


---



---



# Lab 3.1.1 
# *Data Wrangling and Munging with Pandas*

## Part 1: Wrangling Data

The term "data wrangling" is analogous to capturing wild horses and getting them into a fenced area; the horses are data and the fencing is your computer. The more common data wrangling tasks include:

- reading flat files
- reading Excel files
- downloading from web pages
  - csv
  - html
  - json

In [3]:
import numpy as np
import pandas as pd

*It is good practice to display the library version numbers for future reference:*

In [4]:
print('Numpy: ', np.__version__)
print('Pandas: ', pd.__version__)

Numpy:  1.16.4
Pandas:  0.24.2


### CSV Files

Below are three attempts to load the file "bikeshare.csv" into a DataFrame named `bikes`. Why are they wrong?

In [5]:

bikes = pd.read_csv('bikeshare.csv', header = None)
print(bikes.head())



bikes = pd.read_csv('bikeshare.csv', header = 1)
print(bikes.head())



bikes = pd.read_csv('bikeshare.csv', header = 0)
print(bikes.head())

        0           1       2   3     4   5        6        7           8   \
0  instant      dteday  season  yr  mnth  hr  holiday  weekday  workingday   
1        1  2011-01-01       1   0     1   0        0        6           0   
2        2  2011-01-01       1   0     1   1        0        6           0   
3        3  2011-01-01       1   0     1   2        0        6           0   
4        4  2011-01-01       1   0     1   3        0        6           0   

           9     10      11    12         13      14          15   16  
0  weathersit  temp   atemp   hum  windspeed  casual  registered  cnt  
1           1  0.24  0.2879  0.81          0       3          13   16  
2           1  0.22  0.2727   0.8          0       8          32   40  
3           1  0.22  0.2727   0.8          0       5          27   32  
4           1  0.24  0.2879  0.75          0       3          10   13  
   1  2011-01-01  1.1  0  1.2  0.1  0.2  6  0.3  1.3  0.24  0.2879  0.81  \
0  2  2011-01-01    1  

?:
ANSWER: Case 1 treats headings as just another data row. Case 2 treats the 1st data row as the column header. Case 3 gets the header right (row 0), but reads each row as a single column (Nb. the other two make that same mistake). 

Load the file "bikeshare.csv" into a DataFrame named `bikes`, and confirm that it was loaded properly:

In [8]:
import numpy as np
bikes = pd.read_csv('bikeshare.csv' )
print(bikes.head())

   instant      dteday  season  yr  mnth  hr  holiday  weekday  workingday  \
0        1  2011-01-01       1   0     1   0        0        6           0   
1        2  2011-01-01       1   0     1   1        0        6           0   
2        3  2011-01-01       1   0     1   2        0        6           0   
3        4  2011-01-01       1   0     1   3        0        6           0   
4        5  2011-01-01       1   0     1   4        0        6           0   

   weathersit  temp   atemp   hum  windspeed  casual  registered  cnt  
0           1  0.24  0.2879  0.81        0.0       3          13   16  
1           1  0.22  0.2727  0.80        0.0       8          32   40  
2           1  0.22  0.2727  0.80        0.0       5          27   32  
3           1  0.24  0.2879  0.75        0.0       3          10   13  
4           1  0.24  0.2879  0.75        0.0       0           1    1  


In [37]:
df = pd.read_excel('Iris.xls', sheet_name = 'Data')


Species_No  Petal_width  Petal_length  Sepal_width  Sepal_length  \
0             1          0.2           1.4          3.5           5.1   
1             1          0.2           1.4          3.0           4.9   
2             1          0.2           1.3          3.2           4.7   
3             1          0.2           1.5          3.1           4.6   
4             1          0.2           1.4          3.6           5.0   
5             1          0.4           1.7          3.9           5.4   
6             1          0.3           1.4          3.4           4.6   
7             1          0.2           1.5          3.4           5.0   
8             1          0.2           1.4          2.9           4.4   
9             1          0.1           1.5          3.1           4.9   
10            1          0.2           1.5          3.7           5.4   
11            1          0.2           1.6          3.4           4.8   
12            1          0.1           1.4          3.0           4.8   
13            1          0.1           1.1          3.0           4.3   
14            1          0.2           1.2          4.0           5.8   
15            1          0.4           1.5          4.4           5.7   
16            1          0.4           1.3          3.9           5.4   
17            1          0.3           1.4          3.5           5.1   
18            1          0.3           1.7          3.8           5.7   
19            1          0.3           1.5          3.8           5.1   
20            1          0.2           1.7          3.4           5.4   
21            1          0.4           1.5          3.7           5.1   
22            1          0.2           1.0          3.6           4.6   
23            1          0.5           1.7          3.3           5.1   
24            1          0.2           1.9          3.4           4.8   
25            1          0.2           1.6          3.0           5.0   
26            1          0.4           1.6          3.4           5.0   
27            1          0.2           1.5          3.5           5.2   
28            1          0.2           1.4          3.4           5.2   
29            1          0.2           1.6          3.2           4.7   
..          ...          ...           ...          ...           ...   
120           3          2.3           5.7          3.2           6.9   
121           3          2.0           4.9          2.8           5.6   
122           3          2.0           6.7          2.8           7.7   
123           3          1.8           4.9          2.7           6.3   
124           3          2.1           5.7          3.3           6.7   
125           3          1.8           6.0          3.2           7.2   
126           3          1.8           4.8          2.8           6.2   
127           3          1.8           4.9          3.0           6.1   
128           3          2.1           5.6          2.8           6.4   
129           3          1.6           5.8          3.0           7.2   
130           3          1.9           6.1          2.8           7.4   
131           3          2.0           6.4          3.8           7.9   
132           3          2.2           5.6          2.8           6.4   
133           3          1.5           5.1          2.8           6.3   
134           3          1.4           5.6          2.6           6.1   
135           3          2.3           6.1          3.0           7.7   
136           3          2.4           5.6          3.4           6.3   
137           3          1.8           5.5          3.1           6.4   
138           3          1.8           4.8          3.0           6.0   
139           3          2.1           5.4          3.1           6.9   
140           3          2.4           5.6          3.1           6.7   
141           3          2.3           5.1          3.1           6.9   
142           3          1.9           5.1          2.7        

Note that we could have used `read.csv()` above. When is `read_table()` necessary?

?:
ANSWER: When `sep` is not the comma character, or we need fine control that `read.csv()` does not provide.

Flat files can be full of surprises. Here are some issues to watch out for:

- separator character is something other than the comma
  - ";", "|", and tab are popular
- newline character is something other than what the O/S expects 
  - Tip: Don't hard-code the character codes for carriage returns, linefeeds, etc. Use Python's built-in representation instead (e.g. Python translates "\n" to the newline character and "\t" to the tab character on any O/S).
- truncated lines
  - if there are empty fields at the end of a line it is possible that their separators will be missing, resulting in a "jagged" file
- embedded commas or quotes
  - a free-text field containing embedded commas may split into separate fields on input
  - a free-text field containing embedded quotes may not parse correctly
- unescaped characters
  - the "\" character indicates a control code to Python, which will break the I/O
    - e.g. the substring "\u0123" will be interpreted as Unicode(0123) -- which may not be what the file creator intended
  - these may need to be fixed by loading whole strings and then parsing into a new data frame
  
Tip: Most issues can be delth with by correctly specifying the parameters of the function you use to load the file. Read the doco before reading the data!

### Reading Excel Files

In [14]:
from pandas import ExcelFile  # Nb. Need to install xlrd from conda (it does not automatically install with pandas)

In [35]:
df = pd.read_excel('Iris.xls', sheet_name = 'Data')
df

Species_No  Petal_width  Petal_length  Sepal_width  Sepal_length  \
0             1          0.2           1.4          3.5           5.1   
1             1          0.2           1.4          3.0           4.9   
2             1          0.2           1.3          3.2           4.7   
3             1          0.2           1.5          3.1           4.6   
4             1          0.2           1.4          3.6           5.0   
5             1          0.4           1.7          3.9           5.4   
6             1          0.3           1.4          3.4           4.6   
7             1          0.2           1.5          3.4           5.0   
8             1          0.2           1.4          2.9           4.4   
9             1          0.1           1.5          3.1           4.9   
10            1          0.2           1.5          3.7           5.4   
11            1          0.2           1.6          3.4           4.8   
12            1          0.1           1.4          3.0           4.8   
13            1          0.1           1.1          3.0           4.3   
14            1          0.2           1.2          4.0           5.8   
15            1          0.4           1.5          4.4           5.7   
16            1          0.4           1.3          3.9           5.4   
17            1          0.3           1.4          3.5           5.1   
18            1          0.3           1.7          3.8           5.7   
19            1          0.3           1.5          3.8           5.1   
20            1          0.2           1.7          3.4           5.4   
21            1          0.4           1.5          3.7           5.1   
22            1          0.2           1.0          3.6           4.6   
23            1          0.5           1.7          3.3           5.1   
24            1          0.2           1.9          3.4           4.8   
25            1          0.2           1.6          3.0           5.0   
26            1          0.4           1.6          3.4           5.0   
27            1          0.2           1.5          3.5           5.2   
28            1          0.2           1.4          3.4           5.2   
29            1          0.2           1.6          3.2           4.7   
..          ...          ...           ...          ...           ...   
120           3          2.3           5.7          3.2           6.9   
121           3          2.0           4.9          2.8           5.6   
122           3          2.0           6.7          2.8           7.7   
123           3          1.8           4.9          2.7           6.3   
124           3          2.1           5.7          3.3           6.7   
125           3          1.8           6.0          3.2           7.2   
126           3          1.8           4.8          2.8           6.2   
127           3          1.8           4.9          3.0           6.1   
128           3          2.1           5.6          2.8           6.4   
129           3          1.6           5.8          3.0           7.2   
130           3          1.9           6.1          2.8           7.4   
131           3          2.0           6.4          3.8           7.9   
132           3          2.2           5.6          2.8           6.4   
133           3          1.5           5.1          2.8           6.3   
134           3          1.4           5.6          2.6           6.1   
135           3          2.3           6.1          3.0           7.7   
136           3          2.4           5.6          3.4           6.3   
137           3          1.8           5.5          3.1           6.4   
138           3          1.8           4.8          3.0           6.0   
139           3          2.1           5.4          3.1           6.9   
140           3          2.4           5.6          3.1           6.7   
141           3          2.3           5.1          3.1           6.9   
142           3          1.9           5.1          2.7        

In [14]:
df = pd.read_excel('Iris.xls', sheet_name = 'Data', header = 0, usecols = [0, 1, 2, 3, 4])

df

Species_No  Petal_width  Petal_length  Sepal_width  Sepal_length
0             1          0.2           1.4          3.5           5.1
1             1          0.2           1.4          3.0           4.9
2             1          0.2           1.3          3.2           4.7
3             1          0.2           1.5          3.1           4.6
4             1          0.2           1.4          3.6           5.0
5             1          0.4           1.7          3.9           5.4
6             1          0.3           1.4          3.4           4.6
7             1          0.2           1.5          3.4           5.0
8             1          0.2           1.4          2.9           4.4
9             1          0.1           1.5          3.1           4.9
10            1          0.2           1.5          3.7           5.4
11            1          0.2           1.6          3.4           4.8
12            1          0.1           1.4          3.0           4.8
13            1          0.1           1.1          3.0           4.3
14            1          0.2           1.2          4.0           5.8
15            1          0.4           1.5          4.4           5.7
16            1          0.4           1.3          3.9           5.4
17            1          0.3           1.4          3.5           5.1
18            1          0.3           1.7          3.8           5.7
19            1          0.3           1.5          3.8           5.1
20            1          0.2           1.7          3.4           5.4
21            1          0.4           1.5          3.7           5.1
22            1          0.2           1.0          3.6           4.6
23            1          0.5           1.7          3.3           5.1
24            1          0.2           1.9          3.4           4.8
25            1          0.2           1.6          3.0           5.0
26            1          0.4           1.6          3.4           5.0
27            1          0.2           1.5          3.5           5.2
28            1          0.2           1.4          3.4           5.2
29            1          0.2           1.6          3.2           4.7
..          ...          ...           ...          ...           ...
120           3          2.3           5.7          3.2           6.9
121           3          2.0           4.9          2.8           5.6
122           3          2.0           6.7          2.8           7.7
123           3          1.8           4.9          2.7           6.3
124           3          2.1           5.7          3.3           6.7
125           3          1.8           6.0          3.2           7.2
126           3          1.8           4.8          2.8           6.2
127           3          1.8           4.9          3.0           6.1
128           3          2.1           5.6          2.8           6.4
129           3          1.6           5.8          3.0           7.2
130           3          1.9           6.1          2.8           7.4
131           3          2.0           6.4          3.8           7.9
132           3          2.2           5.6          2.8           6.4
133           3          1.5           5.1          2.8           6.3
134           3          1.4           5.6          2.6           6.1
135           3          2.3           6.1          3.0           7.7
136           3          2.4           5.6          3.4           6.3
137           3          1.8           5.5          3.1           6.4
138           3          1.8           4.8          3.0           6.0
139           3          2.1           5.4          3.1           6.9
140           3          2.4           5.6          3.1           6.7
141           3          2.3           5.1          3.1           6.9
142           3          1.9           5.1          2.7           5.8
143           3          2.3           5.9          3.2           6.8
144           3          2.5           5.7          3.3           6.7
145           3

So, this file appears to have an embedded table of aggregates on the same sheet as the raw data (a naughty but common practice amongst analysts).

It is usually better to load data correctly than to meddle with the source file or load it 'warts and all' and then try to parse it in code. The Pandas functions for reading files have parameters that provide the control we need. For ecxample, we could make multiple calls to `read_excel()`, using combinations of the `header`, `usecols`, `skiprows`, `nrows`, and `skipfooter` parameters to load one table at a time from a spreadsheet with multiple tables.

Load the above file without the unwanted columns:

In [15]:
df = pd.read_excel('Iris.xls', sheet_name = 'Data', header = 0, usecols = [0, 1, 2, 3, 4])

df

Species_No  Petal_width  Petal_length  Sepal_width  Sepal_length
0             1          0.2           1.4          3.5           5.1
1             1          0.2           1.4          3.0           4.9
2             1          0.2           1.3          3.2           4.7
3             1          0.2           1.5          3.1           4.6
4             1          0.2           1.4          3.6           5.0
5             1          0.4           1.7          3.9           5.4
6             1          0.3           1.4          3.4           4.6
7             1          0.2           1.5          3.4           5.0
8             1          0.2           1.4          2.9           4.4
9             1          0.1           1.5          3.1           4.9
10            1          0.2           1.5          3.7           5.4
11            1          0.2           1.6          3.4           4.8
12            1          0.1           1.4          3.0           4.8
13            1          0.1           1.1          3.0           4.3
14            1          0.2           1.2          4.0           5.8
15            1          0.4           1.5          4.4           5.7
16            1          0.4           1.3          3.9           5.4
17            1          0.3           1.4          3.5           5.1
18            1          0.3           1.7          3.8           5.7
19            1          0.3           1.5          3.8           5.1
20            1          0.2           1.7          3.4           5.4
21            1          0.4           1.5          3.7           5.1
22            1          0.2           1.0          3.6           4.6
23            1          0.5           1.7          3.3           5.1
24            1          0.2           1.9          3.4           4.8
25            1          0.2           1.6          3.0           5.0
26            1          0.4           1.6          3.4           5.0
27            1          0.2           1.5          3.5           5.2
28            1          0.2           1.4          3.4           5.2
29            1          0.2           1.6          3.2           4.7
..          ...          ...           ...          ...           ...
120           3          2.3           5.7          3.2           6.9
121           3          2.0           4.9          2.8           5.6
122           3          2.0           6.7          2.8           7.7
123           3          1.8           4.9          2.7           6.3
124           3          2.1           5.7          3.3           6.7
125           3          1.8           6.0          3.2           7.2
126           3          1.8           4.8          2.8           6.2
127           3          1.8           4.9          3.0           6.1
128           3          2.1           5.6          2.8           6.4
129           3          1.6           5.8          3.0           7.2
130           3          1.9           6.1          2.8           7.4
131           3          2.0           6.4          3.8           7.9
132           3          2.2           5.6          2.8           6.4
133           3          1.5           5.1          2.8           6.3
134           3          1.4           5.6          2.6           6.1
135           3          2.3           6.1          3.0           7.7
136           3          2.4           5.6          3.4           6.3
137           3          1.8           5.5          3.1           6.4
138           3          1.8           4.8          3.0           6.0
139           3          2.1           5.4          3.1           6.9
140           3          2.4           5.6          3.1           6.7
141           3          2.3           5.1          3.1           6.9
142           3          1.9           5.1          2.7           5.8
143           3          2.3           5.9          3.2           6.8
144           3          2.5           5.7          3.3           6.7
145           3

### Importing Data Directly from the Web

We usually want to store a local copy of a data file that we download from the Web, but when data retention is not a priority it is convenient to download the data directly into our running Python environment.

#### Importing Text Files from the Web

The web is the 'wild west' of data formats. However, we can usually expect good behaviour from files that are automatically generated by a service, such as the earthquake report:

In [39]:
df = pd.read_csv('https://earthquake.usgs.gov/earthquakes/feed/v1.0/summary/2.5_hour.csv')
df.head()

time   latitude   longitude   depth   mag magType  \
0  2019-10-04T17:54:34.440Z  36.822834 -121.561333    7.33  2.63      md   
1  2019-10-04T17:33:39.910Z  66.315800 -157.402500    0.00  2.70      ml   
2  2019-10-04T17:03:50.620Z  -6.238700  155.199600  167.80  4.90      mb   

    nst   gap     dmin   rms  ...                   updated  \
0  62.0  29.0  0.02804  0.11  ...  2019-10-04T17:59:02.407Z   
1   NaN   NaN      NaN  0.84  ...  2019-10-04T17:44:22.815Z   
2   NaN  89.0  3.64300  0.73  ...  2019-10-04T17:22:19.040Z   

                                   place        type horizontalError  \
0        3km SW of San Juan Bautista, CA  earthquake            0.17   
1              69km SSW of Kobuk, Alaska  earthquake             NaN   
2  32km WNW of Panguna, Papua New Guinea  earthquake           10.30   

  depthError  magError  magNst     status  locationSource magSource  
0       0.33     0.180    66.0  automatic              nc        nc  
1       0.30       NaN     NaN  automatic              ak        ak  
2       6.30     0.036   246.0   reviewed              us        us  

[3 rows x 22 columns]

#### Importing HTML Files from the Web

Working with unstructured HTML files relies heavily on library functions. This one, however, is well-structured:

In [41]:
url = 'http://www.fdic.gov/bank/individual/failed/banklist.html'
df = pd.read_html(url)
df

[                                             Bank Name                City  \
 0                                 The Enloe State Bank              Cooper   
 1                  Washington Federal Bank for Savings             Chicago   
 2      The Farmers and Merchants State Bank of Argonia             Argonia   
 3                                  Fayette County Bank          Saint Elmo   
 4    Guaranty Bank, (d/b/a BestBank in Georgia & Mi...           Milwaukee   
 5                                       First NBC Bank         New Orleans   
 6                                        Proficio Bank  Cottonwood Heights   
 7                        Seaway Bank and Trust Company             Chicago   
 8                               Harvest Community Bank          Pennsville   
 9                                          Allied Bank            Mulberry   
 10                        The Woodbury Banking Company            Woodbury   
 11                              First CornerStone B

#### Importing XML Files from the Web

XML files are semi-structured, but you're at the mercy of the file creator. If every record has the same format it will be much easier, but practical applications often require a lot of custom code. Here is an example that includes a nice parser class: http://www.austintaylor.io/lxml/python/pandas/xml/dataframe/2016/07/08/convert-xml-to-pandas-dataframe/

#### Importing JSON Files from the Web

Like XML, JSON files are semi-structured and may require work to capture the schema into a dataframe. Here is a simple example: 

In [16]:
url = 'https://raw.githubusercontent.com/chrisalbon/simulated_datasets/master/data.json'

# Load the first sheet of the JSON file into a data frame
df = pd.read_json(url, orient = 'columns')
df.head()

integer            datetime  category
0         5 2015-01-01 00:00:00         0
1         5 2015-01-01 00:00:01         0
10        5 2015-01-01 00:00:10         0
11        5 2015-01-01 00:00:11         0
12        8 2015-01-01 00:00:12         0

## Part 2: Data Munging

Data munging is manipulating data to get it into a form that we can start running analyses on (which usually means getting the data into a DataFrame). Before we get to this stage, we may need to remove headers or footers, transpose columns to rows, split wide data tables into long ones, and so on. (Nb. Excel files can be particularly troublesome, because users can format their data in mixed, complex shapes.) Essentially, we need to follow Hadley Wickham's guidelines for tidy datasets (http://vita.had.co.nz/papers/tidy-data.html):

The end goal of the cleaning data process:

- each variable should be in one column
- each observation should comprise one row
- each type of observational unit should form one table
- include key columns for linking multiple tables
- the top row contains (sensible) variable names
- in general, save data as one file per table


### Dataset Morphology

Once we have our dataset in a DataFrame (or Series, if our data is only 1-dimensional), we can start examining its size and content.

How many rows and columns are in `bikes`?

In [45]:
bikes.shape


(17379, 17)

What are the column names in `bikes`?

In [46]:
bikes.columns


Index(['instant', 'dteday', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday',
       'workingday', 'weathersit', 'temp', 'atemp', 'hum', 'windspeed',
       'casual', 'registered', 'cnt'],
      dtype='object')

What are the data types of these columns?

In [47]:
bikes.dtypes


instant         int64
dteday         object
season          int64
yr              int64
mnth            int64
hr              int64
holiday         int64
weekday         int64
workingday      int64
weathersit      int64
temp          float64
atemp         float64
hum           float64
windspeed     float64
casual          int64
registered      int64
cnt             int64
dtype: object

What is the (row) index for this DataFrame?

In [48]:
bikes.index


RangeIndex(start=0, stop=17379, step=1)

https://www.dataquest.io/blog/python-json-tutorial/

## Slicing and Dicing

It is often preferable to refer to DataFrame columns by name, but there is more than one way to do this. 
Do `bikes['season']` and `bikes[['season']]` give the same object? Demonstrate:

In [ ]:
url='http//www.dataquest.io/blog/python-json-tutorial'
df = pd.read_json(url, orient = 'columns')
df.head()

How would we use object notation to show the first 4 rows of `atemp`?

In [70]:
print(bikes.atemp[:4])


0    0.2879
1    0.2727
2    0.2727
3    0.2879
Name: atemp, dtype: float64


Algorithms that loop over multiple columns often access DataFrame columns by index. However, none of the following work (try them out by uncommenting / removing the "#E: " ): 

In [61]:
#bikes[0]
bikes[0]
#bikes[0,0]
#bikes[[0,0]]

KeyError: 0

What is the correct way to access the 1st row of the DataFrame by its index?

In [26]:


print(bikes.loc[:0])


   instant      dteday  season  yr  mnth  hr  holiday  weekday  workingday  \
0        1  2011-01-01       1   0     1   0        0        6           0   

   weathersit  temp   atemp   hum  windspeed  casual  registered  cnt  
0           1  0.24  0.2879  0.81        0.0       3          13   16  


What is the correct way to access the 2nd column of the DataFrame by its index?

In [22]:
#ANSWER
bikes.iloc[:, 1]

0        2011-01-01
1        2011-01-01
2        2011-01-01
3        2011-01-01
4        2011-01-01
5        2011-01-01
6        2011-01-01
7        2011-01-01
8        2011-01-01
9        2011-01-01
10       2011-01-01
11       2011-01-01
12       2011-01-01
13       2011-01-01
14       2011-01-01
15       2011-01-01
16       2011-01-01
17       2011-01-01
18       2011-01-01
19       2011-01-01
20       2011-01-01
21       2011-01-01
22       2011-01-01
23       2011-01-01
24       2011-01-02
25       2011-01-02
26       2011-01-02
27       2011-01-02
28       2011-01-02
29       2011-01-02
            ...    
17349    2012-12-30
17350    2012-12-30
17351    2012-12-30
17352    2012-12-30
17353    2012-12-30
17354    2012-12-30
17355    2012-12-31
17356    2012-12-31
17357    2012-12-31
17358    2012-12-31
17359    2012-12-31
17360    2012-12-31
17361    2012-12-31
17362    2012-12-31
17363    2012-12-31
17364    2012-12-31
17365    2012-12-31
17366    2012-12-31
17367    2012-12-31


## Handling Missing Values

What is the Pandas `isnull` function for? 

?
ANSWER:

We can apply `isnull` to the `bikes` DataFrame to show the result for every element:

In [82]:
bikes.isnull().head()

instant  dteday  season     yr   mnth     hr  holiday  weekday  workingday  \
0    False   False   False  False  False  False    False    False       False   
1    False   False   False  False  False  False    False    False       False   
2    False   False   False  False  False  False    False    False       False   
3    False   False   False  False  False  False    False    False       False   
4    False   False   False  False  False  False    False    False       False   

   weathersit   temp  atemp    hum  windspeed  casual  registered    cnt  
0       False  False  False  False      False   False       False  False  
1       False  False  False  False      False   False       False  False  
2       False  False  False  False      False   False       False  False  
3       False  False  False  False      False   False       False  False  
4       False  False  False  False      False   False       False  False

However, we usually start at a higher level. How many nulls are in `bikes` altogether?

In [27]:
bikes.isnull().count()


instant       17379
dteday        17379
season        17379
yr            17379
mnth          17379
hr            17379
holiday       17379
weekday       17379
workingday    17379
weathersit    17379
temp          17379
atemp         17379
hum           17379
windspeed     17379
casual        17379
registered    17379
cnt           17379
dtype: int64

If this result were nonzero we would next want to find out which columns contained nulls. How can this be done in one line of code?

In [28]:
bikes.isnull().sum()


instant       0
dteday        0
season        0
yr            0
mnth          0
hr            0
holiday       0
weekday       0
workingday    0
weathersit    0
temp          0
atemp         0
hum           0
windspeed     0
casual        0
registered    0
cnt           0
dtype: int64

What is the Numpy object `nan` used for? (Write a descriptive answer.)

?
ANSWER: Marking a data point as invalid.

Write (and verify) a function that performs scalar division with built-in handling of the edge case (i.e. return a value instead of just trapping the error):

In [0]:
#ANSWER
def divide(dividend, divisor):
    if divisor == 0:
        quotient = np.nan
    else:
        quotient = dividend / divisor
    return (quotient)

print(divide(1, 0))


Apply the Pandas `isna` function to the following data objects:

In [29]:
x = 2.3
y = np.nan
print(x, y)

2.3 nan


In [0]:
2.3 nan


In [60]:
array = np.array([[1, np.nan, 3], [4, 5, np.nan]])
print(array)

[[ 1. nan  3.]
 [ 4.  5. nan]]


How is the pandas I/O parameter `na_values` used?

? ANSWER: 

## Data Profiling

### Counts

When there are categorical variables in a dataset we will want to know how many possible values there are in each column. (Nb. If the dataset is a sample of a larger one, our sample may not capture all possible values of every categorical.)

How many (different) seasons are in `bikes`?

In [31]:
bikes['season'].value_counts()



3    4496
2    4409
1    4242
4    4232
Name: season, dtype: int64

### Ranges

Print the range of the `instant`, `dteday`, and `windspeed` columns: 

In [33]:
print('instant:', bikes['instant'].min(), 'to', bikes['instant'].max())
print('dteday:', bikes['dteday'].min(), 'to', bikes['dteday'].max())
print('windspeed:', bikes['windspeed'].min(), 'to', bikes['windspeed'].max())


instant: 1 to 17379
dteday: 2011-01-01 to 2012-12-31
windspeed: 0.0 to 0.8507


Compute and print the overall minimum and maximum of the numeric data columns:

In [59]:
bikes_min, bikes_max = ((min(bikes.min()), (max(bikes.max()))
bikes_min, bikes_max

SyntaxError: invalid syntax (<ipython-input-59-1340f6eecbb4>, line 2)

### Quantiles

Pandas makes computing quantiles easy. This is how to get the median of a Series:

In [39]:
bikes['atemp'].quantile(0.5)

0.4848

Of course, the `quantiles` method can take a tuple as its argument. Compute the 10th, 25th, 50th, 75th, and 90th percentiles in one line of code: 

In [58]:
bikes['atemp'].quantile((0.1, 0.25, 0.5, 0.75, 0.9))

0.10    0.2424
0.25    0.3333
0.50    0.4848
0.75    0.6212
0.90    0.6970
Name: atemp, dtype: float64

### Cuts

Sometimes we want to split the sample not by the quantiles of the distribution but by the range of the data. Let's take a closer look at `atemp`:

In [0]:
type(bikes['atemp'])

In [0]:
bikes.sample(5)

Suppose we decide to sort these values into 4 bins of equal width, but we want to apply the resulting groups to the entire DataFrame. Basically, we need to add a row label that indcates which bin each sample belongs in. Let's call this label "atemp_level", and use the `cut` method to populate it:

In [0]:
atemp_level = pd.cut(bikes['atemp'], bins = 4)         

What is `atemp_level`?

In [0]:
#ANSWER


Here is a random sample of `atemp_level`:

In [0]:
atemp_level.sample(5)          

So, by default, `cut` produces labels that indicate the bin boundaries for each element in the series it was applied to. Usually, we will specify labels that are appropriate to the discretisation we are applying:

In [0]:
atemp_level = pd.cut(bikes['atemp'], bins = 4, labels = ["cool", "mild", "warm", "hot"])
atemp_level.sample(5)          

Incorporate the new `atemp_level` column into the `bikes` DataFrame and use it to count the number of "mild" `atemp` entries in `season` 2:

In [0]:
#ANSWER


*Nb. The `atemp_level` variable we created is what the R language calls a "factor". Pandas has introduced a new data type called "category" that is similar to R's factors.*

# Synthetic Data

Sometimes we may want to generate test data, or we may need to initalise a series, matrix, or data frame for input to an algorithm. Numpy has several methods we can use for this.

Execute the following, then check the shape and content of each variable:

In [41]:
# Creating arrays with initial values
a = np.zeros((3))
b = np.ones((1,3))
c = np.random.randint(1,10,(2,3,4))   # randint(low, high, size)
d = np.arange(4)
e = np.array( [[1,2,3,4], [5,6,7,8]] )


In [44]:
a = np.zeros((3))
print(a)

[0. 0. 0.]


In [45]:
print(b)
b = np.ones((1,3))

[[1. 1. 1.]]


In [46]:
c = np.random.randint(1,10,(2,3,4))
print(c)

[[[2 1 6 7]
  [3 2 7 9]
  [4 8 1 4]]

 [[7 9 3 7]
  [5 7 6 3]
  [3 2 7 6]]]


In [48]:
d = np.arange(4)
print(d)

[0 1 2 3]


In [51]:
e = np.array( [[1,2,3,4], [5,6,7,8]] )
print(e)

[[1 2 3 4]
 [5 6 7 8]]


# Cleaning Data

## Load Data

Load rock.csv and clean the dataset.

## Check Column Names

Check column names and clean.

# Replace Null Values With 0

Check 'release' column whether this column have any null value or not. Replace null value with 0.

# Check Datatypes of Dataset

Check datatypes of the dataset. Is there any column which should be int instead of object? Fix the column. 

# Check Min, Max of Each Column

Is there any illogical value in any column? How can we fix that?

# Write Some Functions

## Write a function that will take a row of a DataFrame and print out the song, artist, and whether or not the release date is < 1970

## Write a function that converts cells in a DataFrame to float and otherwise replaces them with np.nan

## Apply these functions to your dataset

## Describe the new float-only DataFrame.



> 
>
>




>


>


>




---



---

> > > > > > > >   © 2019 Data Science Institute of Australia


---



---

